flux-middle/src/cache.rs

```diff
@@ -0,0 +1,55 @@
+use std::{fs::File, path::PathBuf};
+
+use flux_common::config::CONFIG;
+use rustc_hash::FxHashMap;
+use rustc_span::def_id::DefId;
+
+pub struct QueryCache {
+    entries: FxHashMap<String, u64>,
+}
+
+impl Default for QueryCache {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl QueryCache {
+    pub fn new() -> Self {
+        QueryCache { entries: FxHashMap::default() }
+    }
+
+    pub fn insert<Tag>(&mut self, did: DefId, constr_hash: u64) {
+        let str = crate::pretty::def_id_to_string(did);
+        self.entries.insert(str, constr_hash);
+    }
+
+    pub fn is_safe<Tag>(&self, did: DefId, constr_hash: u64) -> bool {
+        let str = crate::pretty::def_id_to_string(did);
```

Comment:
This string representation for `DefId` is not guaranteed to be unique, so I wouldn't use it to hash stuff. Why not only store the constraint hash?

Reply:
I was using the `fn` name as a form of "garbage" collection -- namely just store (the most recent) hash for each function, as opposed to every query since the last creation of the file. 

I changed it to use [`def_path_str`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.def_path_str)

which seems stable and unique and "fully qualified" e.g.

* `runtime::<impl types::VmCtx>::translate_path` 
* `wrappers::wasi_path_readlink`

Reply:
ok, that's fair we don't want the cache growing indefinitely. I think we can at some point implement a lru cache or some other form of eviction strategy.